From 90ca8dbf64bbd19b7a23f8f364214c69527ee5b3 Mon Sep 17 00:00:00 2001
From: Pierre Riteau <priteau@uchicago.edu>
Date: Mon, 18 Jun 2018 19:33:53 +0200
Subject: [PATCH 2/2] Allow to specify a different region for the domain admin
 client

In a multi-region setup with a master Keystone and slaves replicating
user tables, Heat must create users in the master region to avoid
triggering replication failures.
---
 heat/common/config.py                                  | 4 ++++
 heat/engine/clients/os/keystone/heat_keystoneclient.py | 6 +++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/heat/common/config.py b/heat/common/config.py
index 024ee8726..d47273fd9 100644
--- a/heat/common/config.py
+++ b/heat/common/config.py
@@ -58,6 +58,10 @@ service_opts = [
                       'SSL is used.')),
     cfg.StrOpt('region_name_for_services',
                help=_('Default region name used to get services endpoints.')),
+    cfg.StrOpt('region_name_for_domain_admin',
+               help=_('Default region name used for operations by the domain '
+                      'admin. It allows to direct user creation to the '
+                      'master Keystone database.')),
     cfg.StrOpt('heat_stack_user_role',
                default="heat_stack_user",
                help=_('Keystone role for heat template-defined users.')),
diff --git a/heat/engine/clients/os/keystone/heat_keystoneclient.py b/heat/engine/clients/os/keystone/heat_keystoneclient.py
index 4f40c0252..d4325b9de 100644
--- a/heat/engine/clients/os/keystone/heat_keystoneclient.py
+++ b/heat/engine/clients/os/keystone/heat_keystoneclient.py
@@ -119,6 +119,10 @@ class KsClientWrapper(object):
             self._client = self._v3_client_init()
         return self._client
 
+    @property
+    def region_name_for_domain_admin(self):
+        return self.context.region_name or cfg.CONF.region_name_for_domain_admin or cfg.CONF.region_name_for_services
+
     @property
     def region_name(self):
         return self.context.region_name or cfg.CONF.region_name_for_services
@@ -153,7 +157,7 @@ class KsClientWrapper(object):
             self._domain_admin_client = kc_v3.Client(
                 session=self.session,
                 auth=self.domain_admin_auth,
-                region_name=self.region_name)
+                region_name=self.region_name_for_domain_admin)
 
         return self._domain_admin_client
 
-- 
2.17.1

